Update usage count in Promotion eligibility check

filippoliverani opened the pull request:
**Description**

This PR addresses issue #3205.

Eligibility check on an Adjustment bound to a Promotion with a usage limit should not include in usage count the current Adjustment itself.

With this PR the eligibility check does not consider the Order of the Adjustment it is applied to.
I updated the check adding an optional array of Orders to exclude from the computation (in the same vein as Promotion `#used_by?`) to avoid any breaking change to Promotion and PromotionCode puplic APIs.

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I have added tests to cover this change

kennyadsl commented:
@filippoliverani Hey there! Can you please rebase against master now that https://github.com/solidusio/solidus/pull/3278 has been merged? It should fix the flaky spec, thanks!

kennyadsl commented on core/app/models/spree/adjustment.rb:
```diff
@@ -56,6 +56,14 @@ class Adjustment < Spree::Base
     extend DisplayMoney
     money_methods :amount
 
+    def self.adjusted_orders(excluded_orders: [])
```
I have a couple of questions about this class method/scope:

1. Maybe a different name could help understand what this method does: what about a simpler `in_completed_orders()`? I think that maybe with `adjusted_orders()` could not be immediately clear that it's just searching into completed orders.
2. I'm not sure that `eligible` should be part of this method. Chaining it from outside adds a bit of duplication but makes the code more flexible IMO, what do you think?

filippoliverani replied:
I agree on both points.
At first I had used a much longer name but it was quite hard to read, `in_completed_orders` strikes a good balance.

filippoliverani commented:
@kennyadsl Thank you for your suggestions, I'll rebase and resolve the problems 👍 .

kennyadsl commented on core/app/models/spree/promotion.rb:
```diff
@@ -166,22 +168,20 @@ def products
     #
     # @return true or false
-    def usage_limit_exceeded?
+    def usage_limit_exceeded?(excluded_orders = [])
```
We need to also update the YARD documentation here.

kennyadsl replied:
Also, can we use a keyword argument here so it's clear, even when this method is called, that the array contains orders to exclude? I feel like this 

```
usage_limit_exceeded?([promotable])
```

could be better for the reader in this form:

```
usage_limit_exceeded?(excluded_orders: [promotable])
```

filippoliverani replied:
> We need to also update the YARD documentation here.

Missed that 😱 



> can we use a keyword argument here so it's clear, even when this method is called, that the array contains orders to exclude?

Absolutely agree, I used a positional argument just to be consistent with `#used_by?` but here I'd prefer a keyword one too.

kennyadsl commented on core/app/models/spree/promotion.rb:
```diff
@@ -166,22 +168,20 @@ def products
-        joins(:order).
-        merge(Spree::Order.complete).
-        distinct.
+    def usage_count(excluded_orders = [])
```
Same for YARD documentation and keyword argument

filippoliverani commented:
Thank you @kennyadsl, I'm still very new to Solidus and your help is much appreciated!

cedum commented on core/app/models/spree/promotion.rb:
```diff
@@ -126,9 +126,10 @@ def activate(order:, line_item: nil, user: nil, path: nil, promotion_code: nil)
       return false if inactive?
-      return false if usage_limit_exceeded?
-      return false if promotion_code && promotion_code.usage_limit_exceeded?
+      return false if usage_limit_exceeded?(excluded_orders: [promotable])
```
I think we must handle here the different cases a `promotable` object can be:
- `Spree::Order`
- `Spree::LineItem`
- `Spree::Shipment`

You can simulate these cases by creating different promos with different [_Action_' adjustment types.](https://cedum.s3-eu-west-2.amazonaws.com/yfzqwsbz4).

filippoliverani replied:
I had this doubt but I thought that I could ignore LineItems and Shipments when computing eligibility check on them because usage count was not directly affected.

Do you think that updating`#eligible?`to handle non Order promotables excluding Orders to which LineItems and Shipments belong to would be enough?

cedum replied:
> Do you think that updating `#eligible?` to handle non Order promotables excluding Orders to which LineItems and Shipments belong to would be enough?

👍 yes. AFAIK, a promotion's usage count is always the number of distinct orders that promo is applied. I cannot think of cases where it's counted by a different object.

cedum commented on core/app/models/spree/adjustment.rb:
```diff
@@ -56,6 +56,17 @@ class Adjustment < Spree::Base
+      joins(:order).
+      merge(Spree::Order.complete).
+      where.not(spree_orders: { id: excluded_orders }).
+      distinct
```
Not strictly related to this PR, since the `distinct` comes from the previous implementation. Just a heads-up.

Last time I checked it, the generated query was very slow because of `DISTINCT(*)` on a relatively big table (~400k rows). It was something like ~1ms query response with `group by` vs ~500ms using `distinct(*)` on my local machine.
It's worth double checking it and, if confirmed, opening a separate issue.

filippoliverani replied:
This could be the right time to speed up the query, I'll profile both versions with some test data, thanks.

filippoliverani replied:
I tested it on about 500k rows of random genereated rows but I couldn't find any big difference. If you have a more realistic dump without sensitive data could you please share it with me?

cedum replied:
🤔 thanks for looking into this. I'll try to create a reproducible case with dummy data then eventually will open a separate issue with steps to reproduce it.

filippoliverani commented:
@cedum Thank you for the feedback, in the meantime I'll check the existing specs too :+1:!

filippoliverani commented:
@cedum now `#eligible?` should handle all promotable types. I updated and refactored a bit its specs to increase coverage.
I saw some other duplicated and maybe obsolete specs in `promotion_spec.rb` but maybe it would be better to clean them up in an other PR,  WDYT?

cedum commented:
> @cedum now `#eligible?` should handle all promotable types. I updated and refactored a bit its specs to increase coverage.

💯LGTM

> I saw some other duplicated and maybe obsolete specs in `promotion_spec.rb` but maybe it would be better to clean them up in an other PR, WDYT?

agree 🤝
